velox/exec/HashTable.cpp

```diff
@@ -1750,18 +1750,63 @@ void HashTable<ignoreNullKeys>::prepareJoinTable(
   }
 }
 
+template <bool ignoreNullKeys>
+inline uint64_t HashTable<ignoreNullKeys>::joinProjectedVarColumnsSize(
+    const std::vector<vector_size_t>& columns,
+    const char* row) const {
+  uint64_t totalBytes{0};
+  for (const auto& column : columns) {
+    if (!rows_->columnTypes()[column]->isFixedWidth()) {
+      totalBytes += rows_->variableSizeAt(row, column);
+    }
+  }
+  return totalBytes;
+}
+
+template <bool ignoreNullKeys>
+inline uint64_t HashTable<ignoreNullKeys>::joinProjectedFixColumnsSize(
+    const std::vector<vector_size_t>& columns) const {
+  uint64_t totalBytes{0};
+  for (const auto& column : columns) {
+    const auto& type = rows_->columnTypes()[column];
+    VELOX_CHECK(type->isFixedWidth());
+    totalBytes += rows_->fixedSizeAt(column);
+  }
+  return totalBytes;
+}
+
 template <bool ignoreNullKeys>
 int32_t HashTable<ignoreNullKeys>::listJoinResults(
+    const std::vector<vector_size_t>& listColumns,
     JoinResultIterator& iter,
     bool includeMisses,
     folly::Range<vector_size_t*> inputRows,
-    folly::Range<char**> hits) {
+    folly::Range<char**> hits,
+    uint64_t maxBytes) {
   VELOX_CHECK_LE(inputRows.size(), hits.size());
-  if (!hasDuplicates_) {
-    return listJoinResultsNoDuplicates(iter, includeMisses, inputRows, hits);
+  std::vector<vector_size_t> varSizeColumns;
+  std::vector<vector_size_t> fixedSizeColumns;
+  for (const auto column : listColumns) {
+    if (rows_->columnTypes()[column]->isFixedWidth()) {
+      fixedSizeColumns.push_back(column);
+    } else {
+      varSizeColumns.push_back(column);
+    }
   }
+  if (varSizeColumns.empty() && !hasDuplicates_) {
```

Comment:
There is another regression here

Reply:
Could you clarify a bit on this part? Is it the row counts limit in fast path?
Maybe we can run some shadow benchmark to see how big of an impact it is to the overall performance?

Reply:
This disable the fast path when string is present.  The overall impact is probably small as hash join is not majority of the computation, but on individual queries this can be bad.

Reply:
You mean listJoinResultsNoDuplicates is fast path compared with duplicate case? Thanks!